### PR TITLE
Xuorig Fix PR - Edge case with GraphQLTypeReference and Schema Transforms

### PR DESCRIPTION
**Pull request by bbakerman**
This is the same code as https://github.com/graphql-java/graphql-java/pull/2905 but with an extra test for `SchemaTransformer` which also fails

### CODE REVIEW

**Comment by bbakerman** [comment]
I debugged this further.  The reason this blows up (even for SchemaTransformer changes along with direct ones) is this

The first step of the schema build code calls this

```
            SchemaUtil.visitPartiallySchema(partiallyBuiltSchema, codeRegistryVisitor, typeCollectingVisitor);
            ImmutableMap<String, GraphQLNamedType> allTypes = typeCollectingVisitor.getResult();
```

At this point the map is meant to contain a map of name -> ACTUAL types (and not type references)

Later the type reference replacement code finds all type references and expects to find the actual type for them in the map above and its does not and asserts.

However this particular edit means that the schema DOES have a field pointer to a actual type but the `GraphQLTypeCollectingVisitor` never grabs it because it only visits types via direct declaration and not via field declaration.

This makes sense in SDL generation because its impossible to create a situation where the actual type is not declared but code declaration can make this happen as demonstrated in this PR test case.

**Comment by bbakerman** [comment]
Ok debugging more - this is more about the Traverser related code

In this "edit case" we have a field `b` that at some point earlier had a type reference to `B` in the field def.  On the original schema build this gets resolved to an actual type.  So now the field reference has `originalType=typeRef("B")` and also `replacedType=actualTypeB`

Then the code does and edit and this gets rid of the direct reference to type B (which came via another field chain) but this field `b` still has `originalType=typeRef("B")` and`replacedType=actualTypeB`.

So Type B is still around in some form

The `graphql.schema.impl.GraphQLTypeCollectingVisitor#GraphQLTypeCollectingVisitor` setup code does this

```
        traverser = new SchemaTraverser(schemaElement -> schemaElement.getChildrenWithTypeReferences().getChildrenAsList());
```

and the callback to the `GraphqlFieldDefinition` is `getChildrenWithTypeReferences` which is code as

```
        return SchemaElementChildrenContainer.newSchemaElementChildrenContainer()
                .child(CHILD_TYPE, originalType)
                .children(CHILD_ARGUMENTS, arguments)
                .children(CHILD_DIRECTIVES, directivesHolder.getDirectives())
                .children(CHILD_APPLIED_DIRECTIVES, directivesHolder.getAppliedDirectives())
                .build();
```

So while the replace type is in place - its never used.

The question is - should it be?

**Comment by bbakerman** [comment]
The code rightly wants a DAG - if we change it, it starts to fail with

`Internal error: should never happen: This schema is not forming an Directed Acyclic Graph`

The challenge here is this.  The edit caused the field to be the only reference to the `type B` BUT the traverser in the schema build uses `getChildrenWithTypeReferences` and this missed this type.

It's as if the specific use case GraphQLTypeCollectingVisitor needs to grab dangling field references and do a fix up if they are missing - eg type referenced but not in the schema in some other way

File: src/main/java/graphql/schema/GraphQLTypeUtil.java [hunk 1]
```diff
@@ -196,12 +196,7 @@ public static <T extends GraphQLType> T unwrapOneAs(GraphQLType type) {
      * @return the underlying type
      */
     public static GraphQLUnmodifiedType unwrapAll(GraphQLType type) {
-        while (true) {
-            if (isNotWrapped(type)) {
-                return (GraphQLUnmodifiedType) type;
-            }
-            type = unwrapOne(type);
-        }
+        return unwrapAllAs(type);
```
Comment (bbakerman) [inline src/main/java/graphql/schema/GraphQLTypeUtil.java:199]: This was probably a mistake to have the well  named unwrapAll return `GraphQLUnmodifiedType` which excludes GraphqlTypeReferences - but API is API so leave it as is

I did change how it works internally here

File: src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java [hunk 2]
```diff
@@ -86,10 +90,14 @@ public TraversalControl visitGraphQLUnionType(GraphQLUnionType node, TraverserCo
 
     @Override
     public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition node, TraverserContext<GraphQLSchemaElement> context) {
-
+        GraphQLNamedType type = unwrapAllAs(node.getType());
+        if (! (type instanceof GraphQLTypeReference)) {
+            fieldActualTypes.put(type.getName(), type);
+        }
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java:96]: It only puts in fields that unwrap to be an actual type

File: src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java [hunk 3]
```diff
@@ -113,17 +121,38 @@ private void assertTypeUniqueness(GraphQLNamedType type, Map<String, GraphQLName
...
+     * missed if it's the only way to get to that type.
+     *
+     * @param visitedTypes the types collected by this visitor
+     *
+     * @return a fixed up map where the only
+     */
+    private Map<String, GraphQLNamedType> fixFieldDanglingTypes(Map<String, GraphQLNamedType> visitedTypes) {
+        for (GraphQLNamedType fieldPointerType : fieldActualTypes.values()) {
+            String typeName = fieldPointerType.getName();
+            visitedTypes.putIfAbsent(typeName, fieldPointerType);
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java:154]: if we have a type pointed to by a field AND it's not the the list of visited types, we add it since the field is the only pointer to it say

File: src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java [hunk 4]
```diff
@@ -112,18 +118,48 @@ private void assertTypeUniqueness(GraphQLNamedType type, Map<String, GraphQLName
...
+        }
+    }
+
+    private void assertUniqueTypeObjects(GraphQLNamedType type, GraphQLType existingType) {
+        // object comparison here is deliberate
+        if (existingType != type) {
+            throw new AssertException(format("All types within a GraphQL schema must have unique names. No two provided types may have the same name.\n" +
+                            "No provided type may have a name which conflicts with any built in types (including Scalar and Introspection types).\n" +
+                            "You have redefined the type '%s' from being a '%s' to a '%s'",
+                    type.getName(), existingType.getClass().getSimpleName(), type.getClass().getSimpleName()));
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java:133]: made a common method here

File: src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy [hunk 5]
```diff
@@ -214,6 +217,124 @@ class GraphQLSchemaTest extends Specification {
...
+        // there are no more actual B object types in the schema tree.
+        def transformed = schema.transform({
+            it.query(schema.queryType.transform({
+                it.replaceFields(fields)
+            }))
+        })
+
+        then:
+        transformed.containsType("B")
+
```
Comment (bbakerman) [inline src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy:257]: Andres original test that was failing

File: src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy [hunk 6]
```diff
@@ -214,6 +217,124 @@ class GraphQLSchemaTest extends Specification {
...
+                    return changeNode(context, newObjectType);
+                }
+                return TraversalControl.CONTINUE
+            }
+        }
+        GraphQLSchema transformed = SchemaTransformer.transformSchema(schema, visitor)
+
+        then:
+        transformed.containsType("B")
```
Comment (bbakerman) [inline src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy:306]: My SchemaTransformer test that was failing and now passing

File: src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy [hunk 7]
```diff
@@ -214,6 +217,124 @@ class GraphQLSchemaTest extends Specification {
...
+        def queryType = newObject().name("Query")
+                .field(newFieldDefinition().name("a").type(typeA))
+                .build()
+
+        when:
+        def schema = GraphQLSchema.newSchema().query(queryType).build()
+        then:
+        schema.getType("A") != null
+        schema.getType("B") != null
+    }
```
Comment (bbakerman) [inline src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy:336]: Direct object test case (no SDL) that was failing and is now passing

File: src/test/groovy/graphql/schema/GraphQLTypeUtilTest.groovy [hunk 8]
```diff
@@ -127,6 +128,38 @@ class GraphQLTypeUtilTest extends Specification {
...
+        typeRef instanceof GraphQLTypeReference
+        (typeRef as GraphQLTypeReference).name == "A"
+
+        when:
+        typeRef = GraphQLTypeUtil.unwrapAllAs(typeRef)
+
+        then:
+        typeRef instanceof GraphQLTypeReference
+        (typeRef as GraphQLTypeReference).name == "A"
+
```
Comment (bbakerman) [inline src/test/groovy/graphql/schema/GraphQLTypeUtilTest.groovy:160]: Extra tests as `unwrapAllAs` was not previously covered

File: src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy [hunk 9]
```diff
@@ -214,6 +217,124 @@ class GraphQLSchemaTest extends Specification {
...
+
+            type B {
+              a: A
+              b: B
+            }
+        '''
+
+        when:
+        def schema = TestUtil.schema(sdl)
+        // When field `a` is filtered out
```
Comment (andimarek) [inline src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy:242]: b is filtered out

Comment (bbakerman) [reply src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy]: updated

File: src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java [hunk 10]
```diff
@@ -36,14 +45,14 @@ public GraphQLTypeCollectingVisitor() {
...
-        return super.visitGraphQLEnumType(node, context);
+        return CONTINUE;
     }
 
     @Override
     public TraversalControl visitGraphQLScalarType(GraphQLScalarType node, TraverserContext<GraphQLSchemaElement> context) {
         assertTypeUniqueness(node, result);
         save(node.getName(), node);
-        return super.visitGraphQLScalarType(node, context);
+        return CONTINUE;
```
Comment (bbakerman) [inline src/main/java/graphql/schema/impl/GraphQLTypeCollectingVisitor.java:55]: all of this super code boiled down to CONTINUE!

**Review by xuorig — COMMENTED** [review]
Appreciate the fix! Very similar solution to the workaround I implemented but much much nicer now that it is built in in the type collector. Thanks 🙇
